sleap/gui/commands.py

```diff
@@ -4062,33 +4076,69 @@ def get_products_of_instances(
+            """
+
             num_instances = len(instances_in_view)
+
             if num_instances < max_num_instances:
                 num_missing = max_num_instances - num_instances
 
                 # Extend the list first
                 instances_in_view.extend([dummy_instance] * num_missing)
 
-            # Permute instances into all possible orderings w/in a view
-            instances_permutations[cam] = permutations(instances_in_view)
+            return instances_in_view
 
-        # Get products of instances from other views into all possible groupings
-        # Ordering of dict_values is preserved in Python 3.7+
-        products_of_instances: Iterator[Iterator[Tuple]] = product(
-            *instances_permutations.values()
-        )
+        # The existing grouping of instances
+        instance_group: Optional[
+            Dict[Camcorder, List[Instance]]
+        ] = session.get_instance_group(frame_idx=frame_idx)
 
-        # Reorganize products by cam and add selected instance to each permutation
         instances_hypotheses: Dict[int, Dict[Camcorder, List[Instance]]] = {}
-        for frame_id, prod in enumerate(products_of_instances):
-            instances_hypotheses[frame_id] = {
-                cam: [*inst] for cam, inst in zip(instances.keys(), prod)
-            }
+        # TODO(LM): This should be skipped if we are doing greedy matching, not if instance_group is None
+        if instance_group is None:
+            # Get permutations of instances from other views
+            instances_permutations: Dict[Camcorder, Iterator[Tuple]] = {}
+            for cam, instances_in_view in instances.items():
+                # Append a dummy instance to all lists of instances if less than the max length
+                instances_in_view = _fill_in_missing_instances(instances_in_view)
+
+                # Permute instances into all possible orderings w/in a view
+                instances_permutations[cam] = permutations(instances_in_view)
+
+            # Get products of instances from other views into all possible groupings
+            # Ordering of dict_values is preserved in Python 3.7+
+            products_of_instances: Iterator[Iterator[Tuple]] = product(
+                *instances_permutations.values()
+            )
+
+            # Reorganize products by cam and add selected instance to each permutation
+            for frame_id, prod in enumerate(products_of_instances):
+                instances_hypotheses[frame_id] = {
+                    cam: list(inst) for cam, inst in zip(instances.keys(), prod)
+                }
+        else:
+            # Remove instances in selected view from instance grouping, we can't assume
+            # that all instances in the selected view will be in the instance_group
+            instance_group.pop(selected_cam, None)
+
+            # Get instances in current view
+            instances_in_view = instances[selected_cam]
```

Comment:
Possible race condition: when we double click an instance to turn it into a user instance, the `triangulateSession` function is called as a callback.
https://github.com/talmolab/sleap/blob/d5d6a43da741304e19771de6009b69f7f9b03063/sleap/gui/app.py#L1232-L1243
But! there isn't enough time for the user instance to be added to the datastructures before we try to get the instances across views, so we get a KeyError on the initial double click.
 ```
 Traceback (most recent call last):
  File "d:\social-leap-estimates-animal-poses\source\sleap\sleap\gui\app.py", line 1243, in _after_plot_update
    self.commands.triangulateSession(cams_to_include=cams_to_include)
  File "d:\social-leap-estimates-animal-poses\source\sleap\sleap\gui\commands.py", line 632, in triangulateSession
    cams_to_include=cams_to_include,
  File "d:\social-leap-estimates-animal-poses\source\sleap\sleap\gui\commands.py", line 244, in execute
    command().execute(context=self, params=kwargs)
  File "d:\social-leap-estimates-animal-poses\source\sleap\sleap\gui\commands.py", line 138, in execute
    okay = self.ask(context, params)
  File "d:\social-leap-estimates-animal-poses\source\sleap\sleap\gui\commands.py", line 3458, in ask
    return cls.verify_views_and_instances(context=context, params=params)
  File "d:\social-leap-estimates-animal-poses\source\sleap\sleap\gui\commands.py", line 3513, in verify_views_and_instances
    show_dialog=show_dialog,
  File "d:\social-leap-estimates-animal-poses\source\sleap\sleap\gui\commands.py", line 3562, in get_and_verify_enough_instances
    cams_to_include=cams_to_include,
  File "d:\social-leap-estimates-animal-poses\source\sleap\sleap\gui\commands.py", line 4130, in get_products_of_instances
    instances_in_view = instances[selected_cam]
KeyError: Camcorder(name=midL, size=[1280, 1024])
 ```
 To reproduce:
 1. double click prediction in view 1 frame 1
 2. Move a node around
 3. double click prediction in view 2 frame 1
 4. Move a node around
 5. double click a prediction in view 3 frame 1
 6. See KeyError
 
### Observations
- Strange that the error does not reproduce unless we move nodes of instances in at least two views.
- The mouse occasionally gets stuck in the grab cursor instead of a normal cursor
- The error is unable to be reproduced using the debugger (due to slowing it down?)